std/array.d

```diff
@@ -778,39 +778,64 @@ private auto arrayAllocImpl(bool minimallyInitialized, T, I...)(I sizes) nothrow
 }
 
 // overlap
-/*
-NOTE: Undocumented for now, overlap does not yet work with ctfe.
+
+
+/++
 Returns the overlapping portion, if any, of two arrays. Unlike $(D
-equal), $(D overlap) only compares the pointers in the ranges, not the
-values referred by them. If $(D r1) and $(D r2) have an overlapping
-slice, returns that slice. Otherwise, returns the null slice.
-*/
-auto overlap(T, U)(T[] r1, U[] r2) @trusted pure nothrow
-if (is(typeof(r1.ptr < r2.ptr) == bool))
+equal), $(D overlap) only compares the pointers and lengths in the
+ranges, not the values referred by them. If $(D r1) and $(D r2) have an
+overlapping slice, returns that slice. Otherwise, returns the null
+slice.
++/
+
+auto overlap(T, U)(T[] a, U[] b) @trusted
+if (is(typeof(a.ptr < b.ptr) == bool))
 {
-    import std.algorithm.comparison : min, max;
-    auto b = max(r1.ptr, r2.ptr);
-    auto e = min(r1.ptr + r1.length, r2.ptr + r2.length);
-    return b < e ? b[0 .. e - b] : null;
+    import std.algorithm.comparison : min;
+
+    alias RetPtr = CommonType!(T*, U*);
+
+    if (a.ptr <= b.ptr && b.ptr < a.ptr + a.length)
```

Comment:
Replace all uses of `a.ptr` or `b.ptr` to `&a[0]` and `&b[0]` respectively. This forces the code to do a bounds check and the previous code is deprecated in `@safe` code.

Reply:
These comparisons are safe on the platforms we support. There has been a long time ago this story of a CPU architecture that would crash the program as soon as an invalid pointer was loaded in a register dedicated to holding pointers. I forgot what machine that was and whether anyone still uses it. I suggest we go with `trusted` here and we defer further engineering to the time when we'll need to support that architecture.